carina-webdriver/src/main/java/com/qaprosoft/carina/core/foundation/webdriver/core/capability/impl/UIAutomator2Capabilities.java

```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public UiAutomator2Options getCapabilities() {
+        UiAutomator2Options options = new UiAutomator2Options();
+        // this step should be executed before getConfigurationCapabilities() to be able to override this capabilities by default appium approach.
+        setLocaleAndLanguage(options);
+        setCapabilitiesSafe(options, getConfigurationCapabilities());
+        return options;
+    }
+
+    @Override
+    public UiAutomator2Options createCapabilitiesFromCustom(Capabilities customCapabilities) {
+        UiAutomator2Options options = new UiAutomator2Options();
+        if (customCapabilities != null) {
+            for (String capabilityName : customCapabilities.getCapabilityNames()) {
+                options.amend(capabilityName, customCapabilities.getCapability(capabilityName));
+            }
+        }
+        return options;
+    }
+
+    @Override
+    public UiAutomator2Options getCapabilitiesWithCustom(Capabilities customCapabilities) {
+        UiAutomator2Options options = getCapabilities();
+        if (customCapabilities != null) {
+            for (String capabilityName : customCapabilities.getCapabilityNames()) {
+                options.amend(capabilityName, customCapabilities.getCapability(capabilityName));
+            }
+        }
+        return options;
+    }
+
+    /*
+     * http://appium.io/docs/en/writing-running-appium/caps/ locale and language
+     * Locale to set for iOS (XCUITest driver only) and Android.
+     * fr_CA format for iOS. CA format (country name abbreviation) for Android
+     */
+    private void setLocaleAndLanguage(UiAutomator2Options options) {
+        // parse locale param as it has language and country by default like en_US
+        String localeValue = Configuration.get(Configuration.Parameter.LOCALE);
+        LOGGER.debug("Default locale value is : " + localeValue);
+        String[] values = localeValue.split("_");
+        if (values.length == 1) {
+            // only locale is present!
+            options.setLocale(localeValue);
+
+            String langValue = R.CONFIG.get("language");
+            if (!langValue.isEmpty()) {
+                LOGGER.debug("Default language value is : " + langValue);
+                // provide extra capability language only if it exists among config parameters...
+                options.setLanguage(langValue);
+            }
+
+        } else if (values.length == 2) {
+                LOGGER.debug("Put language and locale to android capabilities. language: {}; locale: {}", values[0], values[1]);
+                options.setLanguage(values[0]);
+                options.setLocale(values[1]);
+        } else {
+            LOGGER.error("Undefined locale provided (ignoring for mobile capabilitites): {}", localeValue);
+        }
+    }
+
+    private void setCapabilitiesSafe(UiAutomator2Options options, Capabilities capabilities) {
```

Comment:
safe? we have something unsafe?

Reply:
For classes, depends on BaseOptions, we cannot set capability using setCapability, because we need to understand, is it w3c-safe, and if it is safe, add Appium prefix. So we need to use amend method instead. But I just copy this method from class to class, so in some classes we do not needed in this method

Reply:
this method simplified and moved to AbstractCapabilities as setCapabilities